cardano-cli/src/Cardano/CLI/Ping.hs

```diff
@@ -0,0 +1,582 @@
+    encodeVersion (NodeToNodeVersionV1 magic) =
+          CBOR.encodeWord 1
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToNodeVersionV2 magic) =
+          CBOR.encodeWord 2
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToNodeVersionV3 magic) =
+          CBOR.encodeWord 3
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToNodeVersionV4 magic mode) = encodeWithMode 4 magic mode
+    encodeVersion (NodeToNodeVersionV5 magic mode) = encodeWithMode 5 magic mode
+    encodeVersion (NodeToNodeVersionV6 magic mode) = encodeWithMode 6 magic mode
+    encodeVersion (NodeToNodeVersionV7 magic mode) = encodeWithMode 7 magic mode
+    encodeVersion (NodeToNodeVersionV8 magic mode) = encodeWithMode 8 magic mode
+    encodeVersion (NodeToNodeVersionV9 magic mode) = encodeWithMode 9 magic mode
+    encodeVersion (NodeToNodeVersionV10 magic mode) = encodeWithMode 10 magic mode
+
+
+    encodeWithMode :: Word -> Word32 -> Bool -> CBOR.Encoding
+    encodeWithMode vn magic mode =
+          CBOR.encodeWord vn
+      <>  CBOR.encodeListLen 2
+      <>  CBOR.encodeInt (fromIntegral magic)
+      <>  CBOR.encodeBool mode
+
+handshakeReq :: [NodeVersion] -> ByteString
+handshakeReq []       = BL.empty
+handshakeReq versions = CBOR.toLazyByteString $ handshakeReqEnc versions
+
+data HandshakeFailure
+  = UnknownVersionInRsp !Word
+  | UnknownKey !Word
+  | UnknownTag !Word
+  | VersionMissmath ![Word]
+  | DecodeError !Word !String
+  | Refused !Word !String
+  deriving Show
+
+newtype KeepAliveFailure = KeepAliveFailureKey Word deriving Show
+
+keepAliveRspDec :: NodeVersion
+                -> CBOR.Decoder s (Either KeepAliveFailure Word16)
+keepAliveRspDec v | v >= NodeToNodeVersionV7 minBound minBound = do
+  len <- CBOR.decodeListLen
+  key <- CBOR.decodeWord
+  case (len, key) of
+    (2, 1) -> Right <$> CBOR.decodeWord16
+    (_, k) -> return $ Left $ KeepAliveFailureKey k
+keepAliveRspDec _ = do
+  key <- CBOR.decodeWord
+  case key of
+    1 -> Right <$> CBOR.decodeWord16
+    k -> return $ Left $ KeepAliveFailureKey k
+
+handshakeDec :: CBOR.Decoder s (Either HandshakeFailure NodeVersion)
+handshakeDec = do
+  _ <- CBOR.decodeListLen
+  key <- CBOR.decodeWord
+  case key of
+    1 -> decodeVersion
+    2 -> do
+      _ <- CBOR.decodeListLen
+      tag <- CBOR.decodeWord
+      case tag of
+        0 -> do -- VersionMismatch
+          len <- CBOR.decodeListLen
+          x <- replicateM len CBOR.decodeWord
+          return $ Left $ VersionMissmath x
+        1 -> do -- HandshakeDecodeError
+          vn <- CBOR.decodeWord
+          msg <- unpack <$> CBOR.decodeString
+          return $ Left $ DecodeError vn msg
+        2 -> do -- Refused
+          vn <- CBOR.decodeWord
+          msg <- unpack <$> CBOR.decodeString
+          return $ Left $ Refused vn msg
+        _ -> return $ Left $ UnknownTag tag
+
+    k -> return $ Left $ UnknownKey k
```

Comment:
Can we avoid the magic numbers and either use a regular sum type or pattern synonyms?

Reply:
I believe these magic numbers are tags in the protocol.  The code is lifted from `network-mux`.  There a conversation that this code should stay there.

Reply:
https://input-output-rnd.slack.com/archives/C0441DTDH8R/p1669389418618309?thread_ts=1668518354.879509&cid=C0441DTDH8R

Reply:
@newhoggy I think that the codecs should be in `cardano-ping` library.

Reply:
This code is already in `cadano-ping`.